File: Assets/Bundles/Shared/Resources/TS/Network/RemoteFunction.lua
-- Compiled with unity-ts v2.1.0-75
local _NetworkAPI = require("Shared/TS/Network/NetworkAPI")
local NetworkAPI = _NetworkAPI.default
local NetworkChannel = _NetworkAPI.NetworkChannel
-- To prevent collisions with RemoteEvent IDs:
local RF_ID_OFFSET = 1000000
local ID_COUNTER = 0
local RemoteFunctionClient
do
	RemoteFunctionClient = setmetatable({}, {
		__tostring = function()
			return "RemoteFunctionClient"
		end,
	})
	RemoteFunctionClient.__index = RemoteFunctionClient
	function RemoteFunctionClient.new(...)
		local self = setmetatable({}, RemoteFunctionClient)
		return self:constructor(...) or self
	end
	function RemoteFunctionClient:constructor(id)
		self.id = id
		self.listening = false
		self.sendId = 0
		self.yieldingThreads = {}
	end
	function RemoteFunctionClient:FireServer(...)
		local args = { ... }
		if not self.listening then
			self:StartListening()
		end
		local _original = self.sendId
		self.sendId += 1
		local sendId = _original
		local _array = { sendId }
		local _length = #_array
		table.move(args, 1, #args, _length + 1, _array)
		local sendArgs = _array
		local thread = coroutine.running()
		self.yieldingThreads[sendId] = thread
		NetworkAPI.fireServer(self.id, sendArgs, NetworkChannel.Reliable)
		return coroutine.yield()
	end
	function RemoteFunctionClient:StartListening()
		if self.listening then
			return nil
		end
		self.listening = true
		NetworkAPI.connect(self.id, function(sendId, ...)
			local args = { ... }
			local _yieldingThreads = self.yieldingThreads
			local _sendId = sendId
			local thread = _yieldingThreads[_sendId]
			local _yieldingThreads_1 = self.yieldingThreads
			local _sendId_1 = sendId
			_yieldingThreads_1[_sendId_1] = nil
			if thread ~= nil then
				local success, err = coroutine.resume(thread, unpack(args))
				if not success then
					print("NETWORK HANDLER ERROR:\n" .. tostring(err))
				end
			end
		end)
	end
end
local RemoteFunctionServer
do
	RemoteFunctionServer = setmetatable({}, {
		__tostring = function()
			return "RemoteFunctionServer"
		end,
	})
	RemoteFunctionServer.__index = RemoteFunctionServer
	function RemoteFunctionServer.new(...)
		local self = setmetatable({}, RemoteFunctionServer)
		return self:constructor(...) or self
	end
	function RemoteFunctionServer:constructor(id)
		self.id = id
	end
	function RemoteFunctionServer:SetCallback(callback)
		if self.disconnect ~= nil then
			self.disconnect()
		end
		self.disconnect = NetworkAPI.connect(self.id, function(clientId, sendId, ...)
			local args = { ... }
			local res = { callback(clientId, unpack(args)) }
			local _array = { sendId }
			local _length = #_array
			table.move(res, 1, #res, _length + 1, _array)
			local argsReturn = _array
			NetworkAPI.fireClient(self.id, clientId, argsReturn, NetworkChannel.Reliable)
		end)
	end
end
local RemoteFunction
do
	RemoteFunction = setmetatable({}, {
		__tostring = function()
			return "RemoteFunction"
		end,
	})
	RemoteFunction.__index = RemoteFunction
	function RemoteFunction.new(...)
		local self = setmetatable({}, RemoteFunction)
		return self:constructor(...) or self
	end
	function RemoteFunction:constructor()
		local id = ID_COUNTER
		ID_COUNTER += 1
		id += RF_ID_OFFSET
		self.Server = RemoteFunctionServer.new(id)
		self.Client = RemoteFunctionClient.new(id)
	end
end
return {
	RemoteFunction = RemoteFunction,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/@Easy/Core/Core~/src/Shared/Network/RemoteFunction.ts
import { Player } from "Shared/Player/Player";
import NetworkAPI, { NetworkChannel } from "./NetworkAPI";

type RemoteParamsToClient<T> = Parameters<
	T extends unknown[]
		? (player: Player, ...args: T) => void
		: T extends unknown
			? (player: Player, arg: T) => void
			: (player: Player) => void
>;

type RemoteParamsToServer<T> = Parameters<
	T extends unknown[] ? (...args: T) => void : T extends unknown ? (arg: T) => void : () => void
>;

type RemoteParamsToAllClients<T> = RemoteParamsToServer<T>;

type RemoteCallbackFromClient<T> = (player: Player, ...args: RemoteParamsToClient<T>) => void;
type RemoteCallbackFromServer<T> = (...args: RemoteParamsToServer<T>) => void;

type RemoteFunctionReturn<RX> = RX extends [infer A] ? A : RX;
type RemoteFunctionCallback<TX, RX> = (player: Player, ...args: RemoteParamsToServer<TX>) => RemoteFunctionReturn<RX>;

// To prevent collisions with RemoteEvent IDs:
const RF_ID_OFFSET = 1000000;
let ID_COUNTER = 0;

const packageMap = new Map<number, number>();

class RemoteFunctionClient<TX extends unknown[] | unknown, RX extends unknown[] | unknown> {
	private listening = false;
	private sendId = 0;

	private readonly yieldingThreads = new Map<number, thread>();

	constructor(private readonly id: number) {}

	public FireServer(...args: RemoteParamsToServer<TX>): RemoteFunctionReturn<RX> {
		if (!this.listening) {
			this.StartListening();
		}
		const sendId = this.sendId++;
		const sendArgs = [sendId, ...args];
		const thread = coroutine.running();
		this.yieldingThreads.set(sendId, thread);
		NetworkAPI.fireServer(this.id, sendArgs, NetworkChannel.Reliable);
		return coroutine.yield() as unknown as RemoteFunctionReturn<RX>;
	}

	private StartListening() {
		if (this.listening) return;
		this.listening = true;
		NetworkAPI.connect(false, this.id, (sendId: number, ...args: unknown[]) => {
			const thread = this.yieldingThreads.get(sendId);
			this.yieldingThreads.delete(sendId);
			if (thread !== undefined) {
				task.spawn(thread, ...args);
			}
		});
	}
}

class RemoteFunctionServer<TX extends unknown[] | unknown, RX extends unknown[] | unknown> {
	private disconnect: (() => void) | undefined;

	constructor(private readonly id: number) {}

	public SetCallback(callback: RemoteFunctionCallback<TX, RX>) {
		if (this.disconnect !== undefined) {
			this.disconnect();
		}
		this.disconnect = NetworkAPI.connect(true, this.id, (player: Player, sendId: number, ...args: unknown[]) => {
			const res = [(callback as Callback)(player, ...args)];
			const argsReturn = [sendId, ...res];
			NetworkAPI.fireClient(this.id, player, argsReturn, NetworkChannel.Reliable);
		});
	}
}

export class RemoteFunction<TX extends unknown[] | unknown, RX extends unknown[] | unknown> {
	public readonly server: RemoteFunctionServer<TX, RX>;
	public readonly client: RemoteFunctionClient<TX, RX>;

	constructor(packageOffset?: number) {
		let id: number;
		if (packageOffset !== undefined) {
			if (packageMap.has(packageOffset)) {
				id = packageMap.get(packageOffset)! + 1;
			} else {
				id = packageOffset;
			}
			packageMap.set(packageOffset, id);
		} else {
			id = ID_COUNTER++;
		}

		id += RF_ID_OFFSET;
		this.server = new RemoteFunctionServer(id);
		this.client = new RemoteFunctionClient(id);
	}
}
